src/beamlime/logging/loggers.py

```diff
@@ -0,0 +1,127 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+import logging
+import sys
+from typing import Any, Callable, Union
+
+
+def hold_logging(logging_func: Callable) -> Callable:
+    from functools import wraps
+    from inspect import signature
+
+    @wraps(logging_func)
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
+        logging._acquireLock()
+        out = logging_func(*args, **kwargs)
+        logging._releaseLock()
+        return out
+
+    wrapper.__signature__ = signature(logging_func)
+    return wrapper
+
+
+class BeamlimeLogger(logging.Logger):
+    """
+    Logger for beamlime application.
+
+    Notes
+    -----
+    ``_logRecordFactory`` should be ``BeamlimeColorLogRecord``
+    since it is the lowest LogRecord class in the inheritance hierarchy.
+    It may be replaced with another class if we want to add more fields.
+    """
+
+    def __init__(self, name: str, level: Union[int, str] = 0) -> None:
+        from .records import BeamlimeColorLogRecord
+
+        self._logRecordFactory = BeamlimeColorLogRecord
+        super().__init__(name, level)
+
+    def _wrap_stack_info(
+        self, stack_info: bool = False, stacklevel: int = 1
+    ) -> tuple[str, int, str, Union[str, None]]:
+        if logging._srcfile:
```

Comment:
Another protected attribute. Why do you need this?

Reply:
> _srcfile is used when walking the stack to check when we've got the first
> caller stack frame, by skipping frames whose filename is that of this
> module's source. It therefore should contain the filename of this module's
> source file.
> Ordinarily we would use __file__ for this, but frozen modules don't always
> have __file__ set, for some reason (see Issue #21736). Thus, we get the
> filename from a handy code object from a function defined in this module.
> (There's no particular reason for picking addLevelName.)

I updated this part, I procrastinated since I had to rewrite ``findCaller``...